keps/sig-cluster-lifecycle/kubeadm/4656-add-instance-configuration/README.md

```diff
@@ -0,0 +1,374 @@
+
+### Non-Goals
+
+Update the CRI socket annotation on Node object to be the latest.
+
+## Proposal
+
+kubeadm init: If the -cri-socket args is specified, the `-cri-socket` args is preferred and the `kubelet-instace-config.yml` configuration file is generated based on the `-cri-socket` args, if the `-cri-socket` args is not specified, the container runtime ennpoint is automatically detect and the configuration file is uploaded to the global configuration.
+
+kubeadm join:  If the `-cri-socket` args is specified, the `-cri-socket` args is used in preference, and the `kubelet-instace-config.yml `configuration file is generated based on `-cri-socket`, overriding the kubelet configuration downloaded from the global configuration; if the `-cri- socket` args is not specified, the kubelet configuration downloaded from the global configuration will be used.
+
+kubeadm upgrade: `kubeadm upgrade node` will check `kubelet-config-instance.yaml`  only.
+
+### Risks and Mitigations
+
+## Design Details
+
+kubeadm init:
+
+* No longer need to write the container-runtime-endpoint to kubeadm-flags.env.
+* No longer need to add the kubeadm.alpha.kubernetes.io/cri-socket annotation.
+* If the -cri-socket args is specified, the `-cri-socket` args is preferred and the `kubelet-instace-config.yml` configuration file is generated based on the `-cri-socket` args, if the `-cri-socket` args is not specified, the container runtime ennpoint is automatically detect and the configuration file is uploaded to the global configuration.
+
+kubeadm join:
+
+* No longer need to add the kubeadm.alpha.kubernetes.io/cri-socket annotation.
+* If the `-cri-socket` args is specified, the `-cri-socket` args is used in preference, and the `kubelet-instace-config.yml `configuration file is generated based on `-cri-socket`, overriding the kubelet configuration downloaded from the global configuration; if the `-cri- socket` args is not specified, the kubelet configuration downloaded from the global configuration will be used.
+
+kubeadm reset:
+
+* There is no need to do anything, as kubelet-instace-config.yml will be cleaned up with the deletion of ``/var/lib/kubelet.`
+
+kubeadm upgrade apply:
+
+* In v1.31-1.33, when `kubeadm upgrade apply`, we will read the cri annotaion (if there is no annotaion, we will detect container runtime endpoint automatically) and patch it to the global configuration and generate `kubelet-instace-config`. 
+* In v1.34+, `kubeadm upgrade apply` will no longer read cri annotation any more.
+
+kubeadm upgrade node:
+
+* in v1.31, for other nodes, `kubeadm upgrade node` will check if the cri annotation is diffent with the global setting. if cri-socket is different, We automatically detect the container runtime endpoint to generate `kubelet-instace-config`, and override the `ContainerRuntimeEndpoint` field to `kubelet-config`, but it will not be saved to the global configuration.
```

Comment:
i think the plan should be the following instead:
- 1.31 kubeadm upgrade takes the annotations values and starts writing them to instance config.
- 1.32 kubeadm upgrade removes the annotation on node objects.
- 1.32 kubeadm upgrade does nothing.